File: server/src/main/kotlin/com/heerkirov/hedge/server/exceptions/Exceptions.kt
package com.heerkirov.hedge.server.exceptions

import com.heerkirov.hedge.server.utils.tuples.Tuple2
import com.heerkirov.hedge.server.utils.tuples.Tuple3
import com.heerkirov.hedge.server.utils.tuples.t2
import com.heerkirov.hedge.server.utils.tuples.t3

//== 基础错误类型 ==

/**
 * 输入存在错误。
 */
abstract class BadRequestException<INFO>(code: String, message: String, info: INFO) : BaseException<INFO>(400, code, message, info)

/**
 * 要求附带登录认证。
 */
abstract class UnauthorizedException<INFO>(code: String, message: String, info: INFO) : BaseException<INFO>(401, code, message, info)

/**
 * 当前登录认证的权限不足以访问。
 */
abstract class ForbiddenException<INFO>(code: String, message: String, info: INFO) : BaseException<INFO>(403, code, message, info)

/**
 * 找不到当前的主体资源。
 */
abstract class NotFoundException<INFO>(code: String, message: String, info: INFO) : BaseException<INFO>(404, code, message, info)

//== 在此基础上扩展的通用错误类型 ==

/**
 * 表单参数的类型错误。
 * info: string: 字段名
 */
open class ParamTypeError(paramName: String, reason: String) : BadRequestException<String>("PARAM_TYPE_ERROR", "Param '$paramName' $reason", paramName), Unchecked

/**
 * 表单参数的值错误。
 * info: string: 字段名
 */
open class ParamError(paramName: String) : BadRequestException<String>("PARAM_ERROR", "Param '$paramName' has incorrect value.", paramName), Unchecked

/**
 * 表单参数的值空缺，但是业务需要这个值。
 * info: string: 字段名
 */
open class ParamRequired(paramName: String) : BadRequestException<String>("PARAM_REQUIRED", "Param '$paramName' is required.", paramName), Unchecked

/**
 * 表单参数的值已填写，但业务不需要这个值。
 * info: string: 字段名
 */
open class ParamNotRequired(paramName: String) : BadRequestException<String>("PARAM_NOT_REQUIRED", "Param '$paramName' is not required.", paramName), Unchecked

/**
 * 表单参数选取的某种目标资源并不存在，因此业务无法进行。
 * info: (string, V[]): 字段名, 资源对象列表
 */
open class ResourceNotExist<V>(prop: String, value: V) : BadRequestException<Tuple2<String, V>>("NOT_EXIST", "Resource of $prop '$value' is not exist.", t2(prop, value))

/**
 * 表单参数选取的某种目标资源在当前业务中不适用，因此业务无法进行。
 * info: (string, V[]): 字段名, 资源对象列表
 */
open class ResourceNotSuitable<V>(prop: String, value: V) : BadRequestException<Tuple2<String, V>>("NOT_SUITABLE", "Resource of $prop 'value' is not suitable.", t2(prop, value))

/**
 * 表单参数选取的某种目标资源无法唯一确定，存在混淆的可能，因此业务无法进行。
 * info: (string, V, S): 字段名, 目标资源，实际字段列表
 */
open class ResourceNotUnique<V, S>(prop: String, value: V, results: Collection<S>) : BadRequestException<Tuple3<String, V, Collection<S>>>("NOT_UNIQUE", "Resource of $prop 'value' is not unique.", t3(prop, value, results))

/**
 * 表单的某种目标资源已经存在，因此业务无法进行。
 * info: (string, string, V[]): 资源名称, 资源字段名, 资源对象
 */
open class AlreadyExists<V> (resource: String, prop: String, value: V) : BadRequestException<Tuple3<String, String, V>>("ALREADY_EXISTS", "$resource with $prop '$value' is already exists.", t3(resource, prop, value))

/**
 * 当前资源存在某种级联的资源依赖，因此业务无法进行。
 * 多见于删除业务，且目标资源不允许被静默操作的情况，需要此错误提示，并搭配一个强制删除参数。
 * info: string: 资源名称
 */
open class CascadeResourceExists<V>(resource: String, prop: String, value: V) : BadRequestException<Tuple3<String, String, V>>("CASCADE_RESOURCE_EXISTS", "$resource depending on $prop '$value' exists.", t3(resource, prop, value))

/**
 * API的操作或一部分操作，因为某种原因拒绝执行。
 */
open class Reject(message: String): BadRequestException<Nothing?>("REJECT", message, null)

/**
 * 由于服务尚未初始化，API不能被调用。
 */
open class NotInit: BadRequestException<Nothing?>("NOT_INIT", "Server is not initialized.", null)

/**
 * 在headers中没有发现任何token，然而此API需要验证。或者token无法被正确解析。
 */
class NoToken : UnauthorizedException<Nothing?>("NO_TOKEN", "No available token.", null)

/**
 * 使用的token是错误的，无法将其与任何token认证匹配。
 */
class TokenWrong : UnauthorizedException<Nothing?>("TOKEN_WRONG", "Token is incorrect.", null)

/**
 * 使用的password是错误的。
 */
class PasswordWrong : UnauthorizedException<Nothing?>("PASSWORD_WRONG", "Password is incorrect.", null)

/**
 * 此API只能在客户端调用。
 */
class OnlyForClient : ForbiddenException<Nothing?>("ONLY_FOR_CLIENT", "This API can only be called from client.", null)

/**
 * 此token只能由localhost使用。
 */
class RemoteDisabled : ForbiddenException<Nothing?>("REMOTE_DISABLED", "This Token can only be used in localhost.", null)

/**
 * 当前主体资源未找到。
 */
class NotFound(resourceName: String? = null) : NotFoundException<Nothing?>("NOT_FOUND", "${resourceName ?: "Resource"} not found.", null)

File: frontend/src/functions/http-client/exceptions.ts
import { UsefulColors } from "@/constants/ui"

/*
 * 异常机制阐述：
 * 每个API都存在返回异常状态的可能性。因此在response中，首先根据返回状态，将response区分为ok, BusinessError, ConnectionError三类。
 * ConnectionError通常在http client中直接抛出。接下来分类BusinessError。它分为以下几类：
 * - UncheckedException: 虽然会抛出，但不应该处理，应该在表单检查中提前发现错误。这类错误忽略不提。
 * - CheckedException: 需要前端处理的异常。这些异常需要加上类型实现写明在API声明的response中。
 * - CommonException: 包括daemon exception和internal error。虽然绝大多数API都有可能抛，但应该在最开始拦截统一处理。
 */

export type AllException =
    DaemonException |
    UncheckedException |
    CheckedException |
    InternalError |
    UnknownError

export type BasicException = BaseException<number, string, unknown>

export interface BaseException<S extends number, C extends string, I> {
    status: S
    code: C
    message: string
    info: I
}

type BadRequestException<C extends string, I> = BaseException<400, C, I>

type UnauthorizedException<C extends string, I> = BaseException<401, C, I>

type ForbiddenException<C extends string, I> = BaseException<403, C, I>

type NotFoundException<C extends string, I> = BaseException<404, C, I>

type InternalError = BaseException<500, "INTERNAL_ERROR", null>

type UnknownError = BaseException<number, "UNKNOWN_ERROR", null>

type DaemonException =
    NotInit |
    NoToken |
    TokenWrong |
    OnlyForLocal

type UncheckedException =
    ParamTypeError |
    ParamError |
    ParamRequired |
    ParamNotRequired

type CheckedException =
    NotFound |
    Reject |
    PasswordWrong |
    ResourceNotExist<string, unknown> |
    ResourceNotSuitable<string, unknown> |
    ResourceNotUnique<string, unknown, unknown> |
    AlreadyExists<string, string, unknown> |
    CascadeResourceExists<string, string, unknown> |
    FileNotFoundError |
    FileNotReadyError |
    LocationNotAccessibleError |
    IllegalFileExtensionError |
    StorageNotAccessibleError |
    ContentParseError |
    RecursiveParentError |
    CannotGiveColorError |
    IllegalConstraintError<string, string, unknown> |
    ConflictingGroupMembersError |
    InvalidRegexError |
    InvalidRuleIndexError

//== 背景异常：不属于详细业务异常，应该统一处理 ==

type NotInit = BadRequestException<"NOT_INIT", null>
type NoToken = UnauthorizedException<"NO_TOKEN", null>
type TokenWrong = UnauthorizedException<"TOKEN_WRONG", null>
type OnlyForLocal = ForbiddenException<"ONLY_FOR_LOCAL", null>

//== 普通非必须异常：因为业务原因抛出的通用异常。这些异常是能通过先前检查避免的 ==

export type ParamTypeError = BadRequestException<"PARAM_TYPE_ERROR", string>
export type ParamError = BadRequestException<"PARAM_ERROR", string>
export type ParamRequired = BadRequestException<"PARAM_REQUIRED", string>
export type ParamNotRequired = BadRequestException<"PARAM_NOT_REQUIRED", string>

//== 普通异常：因为业务原因抛出的通用异常 ==

export type PasswordWrong = UnauthorizedException<"PASSWORD_WRONG", null>
export type NotFound = NotFoundException<"NOT_FOUND", null>
export type Reject = BadRequestException<"REJECT", null>
export type ResourceNotExist<P extends string, V> = BadRequestException<"NOT_EXIST", [P, V]>
export type ResourceNotSuitable<P extends string, V> = BadRequestException<"NOT_SUITABLE", [P, V]>
export type ResourceNotUnique<P extends string, V, S> = BadRequestException<"NOT_UNIQUE", [P, V, S[]]>
export type AlreadyExists<R extends string, P extends string, V> = BadRequestException<"ALREADY_EXISTS", [R, P, V]>
export type CascadeResourceExists<R extends string, P extends string, V> = BadRequestException<"CASCADE_RESOURCE_EXISTS", [R, P, V]>

//== 扩展异常：因为业务原因抛出的非通用异常，这些异常用在特定的API中描述具体的事项 ==

export type FileNotFoundError = BadRequestException<"FILE_NOT_FOUND", null>
export type FileNotReadyError = BadRequestException<"FILE_NOT_READY", null>
export type LocationNotAccessibleError = BadRequestException<"LOCATION_NOT_ACCESSIBLE", null>
export type IllegalFileExtensionError = BadRequestException<"ILLEGAL_FILE_EXTENSION", string>
export type StorageNotAccessibleError = BadRequestException<"STORAGE_NOT_ACCESSIBLE", null>
export type ContentParseError = BadRequestException<"CONTENT_PARSE_ERROR", string>
export type RecursiveParentError = BadRequestException<"RECURSIVE_PARENT", null>
export type CannotGiveColorError = BadRequestException<"CANNOT_GIVE_COLOR", null>
export type IllegalConstraintError<P extends string, R extends string, V> = BadRequestException<"ILLEGAL_CONSTRAINT", [P, R, V[]]>
export type ConflictingGroupMembersError = BadRequestException<"CONFLICTING_GROUP_MEMBERS", ConflictingMembers[]>
export type InvalidRegexError = BadRequestException<"INVALID_REGEX", string>
export type InvalidRuleIndexError = BadRequestException<"INVALID_RULE_INDEX", [string, string, string]>

interface ConflictingMembers { group: Member, force: boolean, members: Member[] }
interface Member { id: number, name: string, color: UsefulColors | null, isExported: boolean }

